src/main/java/com/github/firmwehr/gentle/firm/optimization/GlobalValueNumbering.java

```diff
@@ -0,0 +1,374 @@
+			for (var node : emergencyIntelligenceIncinerator) {
+				LOGGER.debug("killing %s", node);
+				Graph.killNode(node);
+			}
+			firm.bindings.binding_irgopt.remove_bads(graph.ptr);
+		}
+
+		/**
+		 * Called when DFS enters block.
+		 *
+		 * @param block The block we just entered.
+		 */
+		private void onEnter(Block block) {
+			openBlocks.add(block);
+
+			var nodes = blocks.get(block);
+
+			// stabalize local block
+			var lastAvailable = new HashMap<>(availableExpressions);
+			var currentAvailable = new HashMap<NodeHashKey, Node>();
+
+			boolean runAgain = true;
+			boolean firstRun = true;
+			while (runAgain) {
+				runAgain = false;
+
+				// if we clear it at the end of the loop, we can't carry over available nodes
+				currentAvailable.clear();
+
+				var replaced = new HashSet<Node>();
+				for (var node : nodes) {
+
+					// if we have been replaced, we don't need to reroute inputs and don't carry over
+					var replacement = lastAvailable.get(new NodeHashKey(node));
+					if (replacement != null && !node.equals(replacement)) {
+						replaced.add(node);
+						continue;
+					}
+
+					// check if we need to rewire outgoing edges
+					if (rewireNode(lastAvailable, node)) {
+						runAgain = true;
+						hasModifiedGraph = true;
+					}
+
+					// rewire may create identical node with already available expression, but we deal with this later
+					currentAvailable.put(new NodeHashKey(node), node);
+				}
+
+				// remove nodes from block if they have been replaced and should no longer be considere
+				nodes.removeAll(replaced);
+				emergencyIntelligenceIncinerator.addAll(replaced);
+
+				// rewire may have created duplicates with available expression, so we always override them
+				lastAvailable = new HashMap<>(currentAvailable);
+				lastAvailable.putAll(availableExpressions);
+
+				// dump graphs if we had actual changes (next statement will destroy this information)
+				if (runAgain) {
+					dumpGraph(block.getGraph(), "gvn-iter");
+				}
+
+				// first run does have access to full block local node identities, so we always need a second run
+				if (firstRun) {
+					runAgain = true;
+					firstRun = false;
+				}
+			}
+
+			// local block stabilized, update available expression
+			availableExpressions.putAll(currentAvailable);
+
+			// TODO: this can lead to two phi nodes becoming identities but can't easily be fixed, either rerun or
+			// TODO: restart gvn for current path from root
+			// check if phi from previous blocks need to be rewired (happens if in-edge was replaced in current block)
+			for (var node : availableExpressions.values()) {
+				if (node instanceof Phi phi) {
+					hasModifiedGraph |= rewireNode(availableExpressions, phi);
+				}
+			}
+
+
+		}
+
+		private boolean rewireNode(HashMap<NodeHashKey, Node> lastAvailable, Node node) {
+			boolean runAgain = false;
+			boolean isPhi = node.getOpCode() == binding_irnode.ir_opcode.iro_Phi;
```

Comment:
Is there a reason you mix opcodes and instanceof checks?

Reply:
The rationale was `instanceof` for readability and opcode comparisons for speed in tight loops. But since nodes don't cache their opcode, instanceof ist Probably fast in all occasions.